### ISSUE TITLE
Tensor normalization at runtime

### ISSUE THREAD

**Issue by lkuich** [Unity-Technologies/ml-agents#2351]
Hi, I have a frozen Tensorflow Image Classification model, and when running a Texture through it with TensorflowSharp, I would resize it, then normalize it like so:
```csharp
// Running previously resized Texture2D
public static TFTensor ToTensor(this Texture2D tex)
{
    var pic = tex.GetPixels32();
    Object.Destroy(tex);

    const int INPUT_SIZE = 128;
    const int IMAGE_MEAN = 0;
    const float IMAGE_STD = 255;

    float[] floatValues = new float[(INPUT_SIZE * INPUT_SIZE) * 3];

    for (int i = 0; i < pic.Length; i++)
    {
        var color = pic[i];

        floatValues[i * 3] = (color.r - IMAGE_MEAN) / IMAGE_STD;
        floatValues[i * 3 + 1] = (color.g - IMAGE_MEAN) / IMAGE_STD;
        floatValues[i * 3 + 2] = (color.b - IMAGE_MEAN) / IMAGE_STD;
    }

    TFShape shape = new TFShape(1, INPUT_SIZE, INPUT_SIZE, 3);
    return TFTensor.FromBuffer(shape, floatValues, 0, floatValues.Length);
}
```

Or in my python implementation:
```python
def __read_tensor_from_bytes(self, image_bytes):
    image_reader = tf.image.decode_image(image_bytes, channels=3)
    float_caster = tf.cast(image_reader, tf.float32)

    dims_expander = tf.expand_dims(float_caster, 0)
    resized = tf.image.resize_bilinear(dims_expander, [self.input_height, self.input_width])
    
    # Normalized here:
    normalized = tf.divide(tf.subtract(resized, [self.input_mean]), [self.input_std])
    sess = tf.Session()
    result = sess.run(normalized)

    return result
```

Now I'm upgrading to Barracuda (conversion went fine), and I can run my model fine, but my accuracy is off because I cannot figure out how to normalize my tensor, like I'm used to doing above.

```csharp
public static Tensor TensorFromTexture(Texture2D image, ImageDims dims)
{
    var imageTensor = new Tensor(image);
    imageTensor = imageTensor.Reshape(
        new TensorShape(1, dims.InputSize, dims.InputSize, 3)
    );
    // TODO: *shrug*
    // var normalized = (imageTensor - dims.ImageMean) / dims.ImageStd;
}
```

Any ideas of how I can do this in Barracuda? I may be confused about the normalization process altogether. The mean and std values depending on the base model seem kind of arbitrary to me. Is this process necessary at runtime, or can I do some work to my training script to "pre-normalize" my inputs?

Thank you

**Comment by harperj** [comment]
Hi @lkuich -- you could use `imageTensor.Data` to get access to the values as a float array.  Then you can use any normal C# logic you would like.

**Comment by vincentpierre** [comment]
Thank you for the discussion. We are closing this issue due to inactivity.
